encrypting pod to router traffic (#160)
skupperproject/skupper-docs: modules/kubernetes/pages/service-certs.adoc

// Type: assembly
[id="encrypting-traffic-pod-router"] 
= Encrypting traffic from a pod to the Skupper router

This section describes how to apply certificates to encrypt the traffic within a cluster.

By default:

* {Skupper-name} creates certificates to establish links between sites using mutual TLS.
This ensures that traffic between sites is encrypted. To use custom certificates for traffic between sites, see xref:index.adoc[].

* Network traffic between pods and the Skupper router is not encrypted. 

You can use certificates to encrypt traffic between pods and the Skupper router as follows:

> * xref:skupper-generated-certs[] - does not require user to provide certificates.

Reviewer: Does not require the user to provided certificates, but the user has to configure them in the deployments.

Author: adding that info in note below the two links, wdyt?

Reviewer: Perfect, thanks

> * xref:user-provided-certs[] - requires that the certificate authority is distributed across all sites.

Reviewer: requires that the CA is distributed across all sites and the server certs signed with that CA stored in a secret also distributed across all sites.

Author: added shorter version (wdyt?)

Reviewer: Perfect, thanks

NOTE: With both procedures, the certificates are provided when you expose the service on the {service-network}.

// Type: procedure
[id="skupper-generated-certs"] 
= Exposing services on the {service-network} with Skupper-generated certificates

> A {skupper-name} installation includes a certificate authority which can generate certificates that can be used to encrypt traffic from the pod to the Skupper router.

Reviewer: the skupper CA is stored in a secret called `skupper-service-client` that will necessary to configure in components included in the VAN that will send request to a service exposed over TLS.

Author: do we need that detail here?

Reviewer: Maybe not here, but it would be helpful to mention it in the same place where `skupper-tls-<service-name>` is; because the user has to be aware of those names to modify the deployments with those certs.

This procedure describes how to use those certificates in your {service-network}.

.Prerequisites

* Access to Kubernetes site where the service is exposed
* Access to Kubernetes site where the service is called from
* service sync is enabled on both sites

.Procedure

. Expose the service on the {service-network}:

.. If you use the `expose` option:
+
--
----
$ skupper expose  <target-type> <target-name> --generate-tls-secrets
----
For example, to expose a `backend` deployment using `http2`:
----
$ skupper expose deployment backend --port 8080 --protocol http2 --generate-tls-secrets
----
--

.. If you use `create` and `bind` options:
+
--
----
$ skupper service create <service-name> --generate-tls-secrets
$ skupper service bind <service-name>  <target-type> <target-name>
----
--

. Check that the service is available on another site:
+
--
----
$ skupper service status
Services exposed through Skupper:
╰─ nghttp2tls (http2 port 443)
----

On this site, a secret is created named `skupper-tls-<service-name>`. 
The secret contains the generated certificates under `data/ca.crt`, `data/tls.crt` and `data/tls.key`.
--

> . Configure components that call the exposed service to use the certificates. 

Author: added step to reflect your comment, wdyt?

+
--
For example, modify a deployment to mount the secret in a container.

----
      volumes:
      - name: certs
        secret:
          secretName: skupper-tls-nghttp2tls
----

--

// Type: procedure
[id="user-provided-certs"] 
= Exposing services on the {service-network} with user-provided certificates

You can encrypt traffic from the pod to the Skupper router using certificates provided by a certificate authority.

.Prerequisites

* Access to Kubernetes site where the service is exposed
* Access to Kubernetes site where the service is called from
* Certificate authority access (intermediate certificate is sufficient)

.Procedure

> . Create a secret from the certificate authority, for example:

Author: +1 to your suggestions... I've left the kubectl commands in as examples, but could remove, wdyt?

> . Create a secret from the certificate authority, for example:

Reviewer: I think that the kubectl examples are really useful, but if adding this example:
```
kubectl create secret tls test-trust-ca --cert=test-trust-ca/tls.crt --key=test-trust-ca/tls.key
kubectl patch secret test-trust-ca -p="{\"data\":{\"ca.crt\": \"$(kubectl get secret test-trust-ca -o json -o=jsonpath="{.data.tls\.crt}")\"}}"

kubectl create secret tls test-service-certs --cert=test-service-certs/tls.crt --key=test-service-certs/tls.key 
kubectl patch secret test-service-certs  -p="{\"data\":{\"ca.crt\": \"$(kubectl get secret test-trust-ca -o json -o=jsonpath="{.data.tls\.crt}")\"}}"
```

// Type: assembly
[id="encrypting-traffic-pod-router"] 
= Encrypting traffic from a pod to the Skupper router

This section describes how to apply certificates to encrypt the traffic within a cluster.

By default:

* {Skupper-name} creates certificates to establish links between sites using mutual TLS.
This ensures that traffic between sites is encrypted. To use custom certificates for traffic between sites, see xref:index.adoc[].

* Network traffic between pods and the Skupper router is not encrypted. 

You can use certificates to encrypt traffic between pods and the Skupper router as follows:

> * xref:skupper-generated-certs[] - does not require user to provide certificates.

Reviewer 2: ```suggestion
* xref:skupper-generated-certs[] - does not require that the user provides certificates.
```

Author: `the user' is also a no no

* xref:user-provided-certs[] - requires that the certificate authority, and the signed certificates, are distributed across all sites.

NOTE: With both procedures, you reference the certificates when you expose the service on the {service-network}.
If you do not reference certificates, the traffic between pods and the Skupper router is unencrypted.

// Type: procedure
[id="skupper-generated-certs"]
== Exposing services on the {service-network} with Skupper-generated certificates

A {skupper-name} installation includes a certificate authority which can generate certificates that can be used to encrypt traffic from the pod to the Skupper router.
This procedure describes how to use those certificates in your {service-network}.

.Prerequisites

> * Access to Kubernetes site where the service is exposed

Reviewer 2: ```suggestion
* Access to the Kubernetes site exposing the service
```
Removed PV.

Author: added 'the', but we don't know who exposed the service, so need to leave PV

> * Access to Kubernetes site where the service is called from

Reviewer 2: ```suggestion
* Access to the Kubernetes site calling the service
```

> * service sync is enabled on both sites

Reviewer 2: ```suggestion
* the service sync is enabled on both sites
```

PV?

.Procedure

. Expose the service on the {service-network}:

.. If you use the `expose` option:
+
--
----
$ skupper expose  <target-type> <target-name> --generate-tls-secrets
----
For example, to expose a `backend` deployment using `http2`:
----
$ skupper expose deployment backend --port 8080 --protocol http2 --generate-tls-secrets
----
--

> .. If you use `create` and `bind` options:

Reviewer 2: ```suggestion
.. If you use the `create` and `bind` options:
```

----

On this site, a secret is created named `skupper-tls-<service-name>`. 
The secret contains the generated certificates under `data/ca.crt`, `data/tls.crt` and `data/tls.key`.
--

. Configure components that call the exposed service to use the certificates stored in `skupper-tls-<service-name>`.
+
--
For example, modify a deployment to mount the secret in a container.

----
      volumes:
      - name: certs
        secret:
          secretName: skupper-tls-nghttp2tls
----
--

. Configure the exposed service, that is the component that responds to the request, to use the certificates stored in `skupper-service-client`.
+
--
The `skupper-service-client` secret contains the certificate and private key of the Skupper certificate authority.
For example, modify a deployment to mount the secret in a container.

----
      volumes:
      - name: certs
        secret:
          secretName: skupper-service-client
----
--

// Type: procedure
[id="user-provided-certs"] 
== Exposing services on the {service-network} with user-provided certificates

You can encrypt traffic from the pod to the Skupper router using certificates provided by a certificate authority.

.Prerequisites

> * Access to Kubernetes site where the service is exposed

Reviewer 2: ```suggestion
* Access to the Kubernetes site that exposes the service
```
